Documentation/dash-hld.md

```diff
@@ -0,0 +1,106 @@
+---
+title: SONiC-DASH High Level Design
+description: Describe SONiC-DASH High Level Design
+last update: 01/13/2022
+---
+
+# SONiC-DASH High Level Design (WIP)
+
+> [!WARNING] 
+> **This document is work in progress**
+
+## Table of Contents
+
+- [Overview](#overview)
+- [Requirements](#requirements)
+- [Architecture](#architecture)
+- [High-Level Design](#high-level-design)
+- [Repositories](#repositories)
+- [Configuration and management](#configuration-and-management)
+- [Restrictions / Limitations](#restrictions--limitations)
+- [Glossary](https://github.com/Azure/DASH/wiki/Glossary)
+- [References](#supplementary-documents)
+
+
+## Overview 
+
+This document describes the **SONiC Disaggregated API for SONiC Hosts** (SONiC-DASH) high level design and architecture. SONiC-DASH (DASH for short) is an open source project that will deliver enterprise network performance to critical cloud applications. The ultimate goal is that DASH will have the same success as SONiC for switches and also be widely adopted as a major **Open NOS for Programmable Hardware Technologies**, including [SmartNICs](https://blogs.nvidia.com/blog/2021/10/29/what-is-a-smartnic/), to supercharge a variety of cloud and enterprise applications.
+  
+DASH extends SONiC APIs and a comprehensive set of object models that initially describe Microsoft Azure’s networking services for the cloud.The project enlists cloud and enterprise providers to further extend DASH to meet their specific needs.
+
+See the project [Scenario Milestone and Scoping](SDN-Features-Packet-Transforms.md#scenario-milestone-and-scoping). 
+
+## Requirements
+
+The overall requirement is to **optimize network SMART Programmable Technologies performance**, and **leverage commodity hardware technology** to achieve **10x or even 100x stateful connection performance**.
+- With the help of network hardware vendors, create an open forum that capitalizes on the use of **programmable networking hardware** including SmartNICs, SmartToRs, SmartAppliances. 
+- Optimize **stateful L4** performance and connection scale by 10x or even 100x when compared to implementations that make extensive use of software. As host networking in the cloud is performed at L4 the resulting performance improvements should be truly significant.
+- Microsoft Azure will integrate and deploy DASH solutions to ensure that scale, monitoring, reliability, availability and constant innovation are proven and hardened. It should be noted that innovations for **in-service software upgrades** (ISSU) and **high availability** (HA) are key tenets of the DASH charter.   
+
+
+## Architecture
+
+SONiC is structured into various containers that communicate through multiple logical databases via a shared Redis instance. DASH will make use of the SONiC infrastructure as shown in the figure below.
+
```

Comment:
**SDN Controller**
SAI is the Southbound interface facing the hardware. 
SONiC will be calling SAI.  
SAI is not exposed as gRPC, and is not in gNMI format.  
We need to have a method to call the card remotely; we are going to use gRPC.  
The idea is to create abstraction layer on top of SAI interface (that suppliers have), with an agent translating and driving the goal sate and calling the SAI API to configure the hardware.
The NB API is a way to control the card through the control-plane.  

The SDN controller is the overlay that we build on top of the physical layer of the infrastructure.  From the point of view of the SDN control plane, when a customer creates an operation (ex: VNET creation) in our user interface, the calls come in to the control plane to allocate the resources, placement management, decide on capacity, etc...  Our control plane needs to call the hardware and the software for configuration.  The API calls we will be calling are the 'NorthBound' interface API's for customers to control the behavior.  gNMI is the open format we are extending with additional set of APIs on top of the gNMI to perform actions related to SDN.  We will have the API to work with the abstraction layer for SDN.  We will have a NB interface for creating and deleting the Elastic Network Interface (ENI), create Access Lists (ACLs) and specify the rules, establish the routing and mappings, encapsulate the traffic; all of the APIs will allow us (from the NB) to control this behavior.  There will be an Agent we will develop and publish to the Community (with help) to provide the translation (NB APIs exposed over gRPC to be called from the control plane).  Each DPU via gNMI will expose its own management ID/IP (depending upon availability of IPv4 address space - we may have to optimize).  _In the case of a Smart Switch_, the SDN controller management service will identify the specific card that needs to be programmed, it will 'call' the card via the exposed interface.  No complex logic will run on the switches (switches do not have a top-level view of other/neighboring switches in the infrastructure).  

For High Availability, the SDN controller selects the pair of cards that will have the same goal-state.  It will always be scoping the control-plane state.  For example:  the SDN controller will call gNMI on Card A and Card B to configure ENIs, update mappings for communication, etc..  The only requirement on the card from the HA perspective is for the cards to setup a channel between themselves for flow synchronization.  We will not be replicating flows from the control-plane.  We will leave this open for the supplier to provide the most efficient method to synchronize this.

Reply:
@KrisNey-MSFT and @chrispsommers I incorporated your writing into the Architecture section and created a layered high-level diagram that Kristina sanity-checked. Obviously, all of it can be improved. But I think we are getting there.

Reply:
Making @KrisNey-MSFT and @chrispsommers change in comment to keep a record. I did add the relevant info into the Architectures section in different format. 
## SDN Architecture
The SDN controller is primarily responsible for controlling the DASH overlay services. The traditional SONiC stack is used to manage the underlay (L3 routing) and hardware platform. The SDN controller communicates to a DASH device through a gNMI endpoint served by a new DASH SDN agent running inside a new SONiC DASH container.  Microsoft will spearhead the development of the SDN Agent and solicit Community participation. This container translates SDN configuration modeled in gNMI into SONiC APPL_DB config objects. The SONiC `orchagent` will be enhanced to translate these objects into SAI_DB objects, including the new DASH-specific SAI objects. An enhanced `syncd` will then configure the dataplane using the vendor-specific SAI library. 
The SDN controller controls the overlay built on top of the physical layer of the infrastructure.  From the point of view of the SDN control plane, when a customer creates an operation (ex: VNET creation) in a portal, the controller allocates the resources, placement management, capacity, etc. via the 'NorthBound' interface APIs.
gNMI schema (TODO - provide URL) will be created for managing DASH services: the Elastic Network Interface (ENI), Access Control Lists (ACLs), routing and mappings, encapsulations; etc.  MSFT will spearhead the development of an SDN Agent and solicit Community participation.
### Multiple DPUs in a Device
Each DPU provides a gNMI endpoint for SDN control through a unique IP address. An appliance or smart switch containing multiple DPUs therefore contains multiple gNMI endpoints for SDN control, and the controller treats each DPU as a separate entity. To conserve IPv4 addresses, such an appliance or switch _might_ contain a proxy (NAT) function to map a single IP address:port combination into multiple DPU endpoints with unique IPv4 addresses.  No complex logic will run on the switches (switches do not have a top-level view of other/neighboring switches in the infrastructure). 
## SDN and DPU High-Availability
For High Availability, the SDN controller selects the pair of cards and configures them identically.  The only requirement on the card from the HA perspective is for the cards to setup a channel between themselves for flow synchronization.  The synchronization mechanism is left for vendors to define and implement.  See <link to HA document>.

Reply:
Content included.